emails/tests/views_tests.py

```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
+
+        (info_log,) = info_logs.records
+        assert info_log.msg == "complaint_notification"
+        assert getattr(info_log, "user_match") == "found"
+        assert getattr(info_log, "relay_action") == "auto_block_spam"
+        assert getattr(info_log, "found_in") == "complained_recipients"
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "commonHeaders"
+        assert getattr(err_log, "found_keys") == (
+            "destination,headers,headersTruncated,messageId,source,timestamp"
+        )
+
+    def test_complaint_no_from_header_error_logged(self):
+        """If the mail object is missing, an error is logged"""
+        complaint_msg = deepcopy(self.complaint_msg)
+        del complaint_msg["mail"]["commonHeaders"]["from"]
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with self.assertLogs(ERROR_LOG) as error_logs:
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 200
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "from"
+        assert getattr(err_log, "found_keys") == "subject,to"
+
+    def test_complaint_unknown_complained_recipient_404(self):
+        """If the recipient is unknown but sender is mask, return 404"""
+        complaint_msg = deepcopy(self.complaint_msg)
+        complaint_msg["complaint"]["complainedRecipients"] = [
+            {"emailAddress": "unknown@somewhere.example.com"}
+        ]
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with (
+            self.assertLogs(INFO_LOG) as info_logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 404
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (info_log,) = info_logs.records
+        assert info_log.msg == "complaint_notification"
+        assert getattr(info_log, "user_match") == "found"
+        assert getattr(info_log, "found_in") == "from_header"
+        assert getattr(info_log, "relay_action") == "auto_block_spam"
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_gather_complainers: unknown complainedRecipient"
+
+    def test_complaint_unknown_complained_recipient_two_masks_404(self):
+        """
+        If the recipient is unknown but two senders are masks, return 404
+
+        Also log the weirdness of two senders that match the user.
+        """
```

Comment:
similar suggestion (non-blocking):

```suggestion
        """
        With unknown emailAddress in complainedRecipients, and 2 masks in From:,
        we can still match the user on From:, perform the appropriate action,
        and log an error.
        """
```